### PR TITLE
Add trainer

### PR DESCRIPTION
**Pull request by vbkaisetsu**
This PR adds a trainer.

I checked the behavior of the trainer using BCCWJ and UniDic.
It seems to be generally working correctly, but the handling of unknown words may need to be fixed in the future.

Therefore, I think it appropriate to keep this feature experimental and add the CLI after the stabilization.

### CODE REVIEW

File: vibrato/src/trainer.rs [hunk 1]
```diff
@@ -1,4 +1,473 @@
...
+                feature,
+            )?;
+        }
+
+        writeln!(
+            &mut connector_wtr,
+            "{} {}",
+            merged_model.right_conn_to_left_feats.len(),
+            merged_model.left_conn_to_right_feats.len(),
+        )?;
```
Comment (kampersanda) [inline vibrato/src/trainer.rs:456]: This is just a confirmation.

The definition of the `matrix.def` is counterintuitive: The first column corresponds to **right** ids, and the second column corresponds to **left** ids. No problem in your code?

**Review by kampersanda — COMMENTED** [review]
Thank you for your great contribution!

Could you add several tests for `Trainer` under `tests`?

File: vibrato/src/trainer.rs [hunk 2]
```diff
@@ -1,4 +1,473 @@
+//! Module for training models.
+//!
+//! # Examples
+//!
+//! ```
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer.rs:6]: This is necessary to use `?` in the example.

[the remark this answers is not in the document]
Comment (kampersanda) [reply vibrato/src/trainer.rs:6]: wow😮

File: vibrato/src/dictionary/character.rs [hunk 3]
```diff
@@ -116,6 +116,13 @@ impl CharProperty {
             .map(|id| u32::try_from(id).unwrap())
     }
 
+    #[inline(always)]
+    pub fn cate_string(&self, cate_id: u32) -> Option<&str> {
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/dictionary/character.rs]: fixed

File: vibrato/src/trainer/corpus.rs [hunk 4]
```diff
@@ -17,59 +37,25 @@ pub struct Word {
...
-    tokens: Vec<Word>,
-}
-
-impl Sentence {
-    /// Returns a slice of tokens.
-    #[allow(unused)]
-    pub fn tokens(&self) -> &[Word] {
-        &self.tokens
-    }
+pub struct Example {
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer/corpus.rs:51]: I added the comment to the `sentence` field.

File: vibrato/src/trainer/corpus.rs [hunk 5]
```diff
@@ -191,31 +131,31 @@ EOS
...
-        assert_eq!("行く", sentence1.tokens()[3].surface());
-        assert_eq!("動詞,イク", sentence1.tokens()[3].features());
+        assert_eq!("トスカーナ", sentence1.tokens[0].surface());
+        assert_eq!("名詞,トスカーナ", sentence1.tokens[0].feature());
+        assert_eq!("地方", sentence1.tokens[1].surface());
+        assert_eq!("名詞,チホー", sentence1.tokens[1].feature());
+        assert_eq!("に", sentence1.tokens[2].surface());
+        assert_eq!("助詞,ニ", sentence1.tokens[2].feature());
+        assert_eq!("行く", sentence1.tokens[3].surface());
+        assert_eq!("動詞,イク", sentence1.tokens[3].feature());
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer/corpus.rs:151]: fixed

File: vibrato/src/trainer/corpus.rs [hunk 6]
```diff
@@ -191,31 +131,31 @@ EOS
...
 
-        assert_eq!("火星", sentence2.tokens()[0].surface());
-        assert_eq!("名詞,カセー", sentence2.tokens()[0].features());
-        assert_eq!("猫", sentence2.tokens()[1].surface());
-        assert_eq!("名詞,ネコ", sentence2.tokens()[1].features());
+        assert_eq!("火星", sentence2.tokens[0].surface());
+        assert_eq!("名詞,カセー", sentence2.tokens[0].feature());
+        assert_eq!("猫", sentence2.tokens[1].surface());
+        assert_eq!("名詞,ネコ", sentence2.tokens[1].feature());
     }
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer/corpus.rs:161]: fixed

File: vibrato/src/trainer.rs [hunk 7]
```diff
@@ -1,4 +1,473 @@
...
+
+use crate::common::MAX_SENTENCE_LENGTH;
+
+/// Trainer of morphological analyzer.
+pub struct Trainer {
+    dict: Dictionary,
+    surfaces: Vec<String>,
+    max_grouping_len: Option<u16>,
+    provider: FeatureProvider,
+    label_id_map: HashMap<String, HashMap<char, u32>>,
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer.rs:99]: fixed

File: vibrato/src/trainer.rs (hunk 1 above)
Comment (vbkaisetsu) [reply vibrato/src/trainer.rs:456]: No problem.
I changed the variable names to make this easy to understand: `i` -> `right_conn_id`, `j` -> `left_conn_id`.

File: vibrato/src/trainer.rs [hunk 8]
```diff
@@ -1,4 +1,473 @@
...
+//! let trainer = Trainer::new(config)?
+//!     .regularization_cost(0.01)
+//!     .max_iter(300)
+//!     .num_threads(20);
+//!
+//! // Loads corpus
+//! let corpus_rdr = File::open("src/tests/resources/corpus.txt")?;
+//! let corpus = Corpus::from_reader(corpus_rdr)?;
+//!
+//! // Files to store results
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/trainer.rs:35]: I changed to use `Vec`

**Comment by vbkaisetsu** [comment]
Since it is very difficult to obtain training results analytically, I would like to keep testing simple.
The example code in `trainer.rs` works as a test. Is this not sufficient?

**Review by kampersanda — COMMENTED** [review]
> Since it is very difficult to obtain training results analytically, I would like to keep testing simple.
> The example code in trainer.rs works as a test. Is this not sufficient?

No testing is required for actual trained costs, as it is difficult to test those values.
The formats of output files have been already tested with `Dictionary::from_readers`.

On the one hand, I guess that entries (except costs) in the output lexicon are deterministic.
At least, a test should be done to see if the output lexicon contains the expected number of entries.

**Pull request merged**